### DISCUSSION TITLE
LDAP stops working after "some time"

### DISCUSSION THREAD

**Discussion by dorianim** [directus/directus#9336] [Feature Requests]
Hi,

I ran into an issue with the LDAP integration: After some time, it just stops working with the error "Unexpected Error" on login. In the console, I can see this error message: 
```json
{"errors":[{"message":"Service returned unexpected error","extensions":{"code":"SERVICE_UNAVAILABLE","service":"ldap","message":"00002020: Operation unavailable without authentication"}}]}
```
After a server restart, the issue is resolved for "some time".
I think the cause for this issue is, that directus only binds once to the LDAP server using the configured bind user. This bind seems to expire after some time, which breaks the login.

I think to properly resolve this, we need to bind to the LDAP (with the bind user) not only once, but at every login attempt.
What do you think @aidenfoxx ?

**Comment by aidenfoxx** [comment]
That could be the case. The "ldapjs" documentation isn't great, but I can now see it says `After the reconnect you maybe need to bind again.`.

We could add an event listener to the client `on('connectTimeout')` to re-bind the user. This is hard for me to verify, however, since my longest bind session is still active. 😅 

A temporary solution might be to set an infinite `connectTimeout` until we have it fixed:

`AUTH_LDAP_CLIENT__CONNECT_TIMEOUT=-1`

**Reply by aidenfoxx** [reply]
This might be the solution https://github.com/ldapjs/node-ldapjs/issues/392

**Reply by dorianim** [reply]
Ah, cool! As I understand it, we'd just have to surround this block
https://github.com/directus/directus/blob/f73df5dffdca52aecae0705e08eb0a0550e0bdc4/api/src/auth/drivers/ldap.ts#L67

With a 
```ts
 client.on('connect', function () {
}
```
Right?

**Reply by aidenfoxx** [reply]
It's a little more complicated since we need to handle the initial bind, as well as future binds inside the connect callback. So we have to handle the resolve/reject in a special way. I've got a PR for it though. I'll upload it soon so you can test.

**Reply by dorianim** [reply]
Ok, cool! Thank you :+1:

**Reply by dorianim** [reply]
>  my longest bind session is still active

Interesting... What backend are you using?
I'm using Samba on Ubuntu 18.04.

**Reply by aidenfoxx** [reply]
We're using Active Directory on some custom linux container. Not sure what distro.

Here if the fix if you wanna give it a try https://github.com/directus/directus/pull/9339. No idea if it'll work as expected!!

**Reply by dorianim** [reply]
Nice! I'll try

**Reply by aidenfoxx** [reply]
Seems to work for me. Running locally, I can disconnect my wifi, see the `read ECONNRESET`, reconnect my wifi, and it rebinds the user as expected.

**Reply by dorianim** [reply]
I waited for some time and after the first login attempt, I got the unexpected error but when trying again it worked.

**Reply by dorianim** [reply]
Just disconnecting and reconnecting works fine for me too.

**Reply by dorianim** [reply]
Unfortunately, it always takes about 30mins for the issue to appear, so the testing is really slow ...

**Reply by aidenfoxx** [reply]
Hm... That's weird. It must try doing the reconnect when you try to login. On my system it reconnects as soon as the connection drops. Maybe it's something to do with the `idleTimeout`.

Not sure how we could handle this instance better. Feels like it could be a race condition where login triggers reconnect, but the login expects the connection to be active.

**Reply by dorianim** [reply]
Yes, that is possible. How about reconnecting for each login and closing the connection afterwards? I think this is what most other tools do (eg. Rocket.Chat: https://github.com/RocketChat/Rocket.Chat/blob/f3e7b95800b68a4c32e36529a62ddcd5f96d73c9/server/lib/ldap/Manager.ts#L25 ). I am not sure however, how this could affect performance...

**Reply by aidenfoxx** [reply]
Maybe that would be the best way to handle this. I just assumed it would be simpler and more efficient to keep a single client active, but it seems not.

Frustratingly I can't reproduce the issues you're having. Even on a natural connection timeout on my side, the client reconnects immediately and works as expected.

**Reply by dorianim** [reply]
I don't know a lot about the underlying protocol of LDAP authentication, but to me it looks like something is just invalidating the bind session without closing the connection in the first place. So the connection stays alive, but requests are rejected as unauthorized.

**Reply by dorianim** [reply]
I also looks like the session is renewed with every request. So the error only occurs when there is no login attempt for over ~30 Minutes.

**Reply by aidenfoxx** [reply]
I guess it makes sense that there is some kind of internal timeout in LDAP for bind sessions. Our server probably has an infinite session length, which is why I don't get the issue.

This is a pain to solve for because we use the bind client not just for authentication, but for verifying the user is still valid on `refresh`, so if we have to rebind the system user every time we get a `refresh` call that's gonna be a nightmare for performance.

The nicest thing would be if there was a callback that's triggered when the bind session expires so we can re-bind then.

**Reply by aidenfoxx** [reply]
I think I have a nice solution. Will push it soon.

**Reply by dorianim** [reply]
Cool!
I just waited for another 40 Minutes and now the requests are just timing out...

**Reply by aidenfoxx** [reply]
Yeah, I'm not sure how to fix this nicely. I'll do some thinking about this tonight and let you know what I figure out.

**Reply by dorianim** [reply]
Thank you very much!
